### DISCUSSION TITLE
if I intend to use straight.el, is `(load "~/crafted-emacs/modules/crafted-early-init-config")` still necessary?

### DISCUSSION THREAD

**Discussion by jf** [SystemCrafters/crafted-emacs#417] [Q&A] [answered]
as per subject.

**Comment by jvdydev** [comment, marked as answer]
Hi,

the `crafted-early-init-config` initializes `package.el` and provides some functionality for automatic fetching of package archives (using `package.el`).
Therefor, loading it when using `straight.el` is probably not what you want.
Instead, you may want to load `crafted-package-config` and configure it so `crafted-install-selected-packages` automatically uses `straight.el`.

We also have an example for [using straight.el with Crafted Emacs](https://github.com/SystemCrafters/crafted-emacs/tree/master/examples/straight-el) (see [Examples Doc](https://github.com/SystemCrafters/crafted-emacs/blob/master/docs/crafted-examples.org) for how to run).

Hope this helps :smile:

**Comment by jf** [comment]
yes, thank you! this is amazing and great that you guys thought to provide code for this too. Thank you!

**Reply by jvdydev** [reply]
No problem, most of the examples are provided by (or built in collaboration with) people in the community who use the respective packages :smile:

**Comment by jf** [comment]
sorry if I can add on some more questions (let me know if it's better to create a new discussion):

I see in the 2 files at https://github.com/SystemCrafters/crafted-emacs/tree/master/examples/straight-el the following:
1. an initial line at the top ending in `-*- lexical-binding: t; -*-`. Is this like the modeline feature for vim, and if so, what does this line do?
2. is it really necessary to have `(provide 'early-init)` and `(provide 'init)` at the bottom of `early-init.el` and `init.el`?

**Reply by sthesing** [reply]
Both things are language features of Emacs lisp.

1. No, this has nothing to do with a modeline or such. It activates lexical binding for the module in question. See [Lexical Binding](https://www.gnu.org/software/emacs/manual/html_node/elisp/Lexical-Binding.html) in the Emacs manual for details.
2. The `provide` statements are the difference between scripts that just run for themselves and modules that can be imported into other scripts. If `(provide 'early-init)` was missing, `(require 'early-init)` wouldn't work. So you probably won't need these statements in `init.el`. Whether you need it in `early-init.el` depends on how you load it into your `init.el`.
